File: backends/spring-boot/src/main/java/com/hellotimepro/springboot/service/SecurityService.java
package com.hellotimepro.springboot.service;

import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTVerificationException;
import com.auth0.jwt.exceptions.TokenExpiredException;
import com.hellotimepro.springboot.config.AppProperties;
import com.hellotimepro.springboot.domain.UserEntity;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.util.Base64;
import java.util.Date;
import java.util.Optional;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class SecurityService {
  private final AppProperties props;
  private final BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(10);
  private final Algorithm algorithm;
  private final JWTVerifier verifier;

  public SecurityService(AppProperties props) {
    this.props = props;
    this.algorithm = Algorithm.HMAC256(props.getJwtSecret());
    this.verifier = JWT.require(algorithm).build();
  }

  public String hashPassword(String plain) {
    return encoder.encode(plain);
  }

  public boolean verifyPassword(String plain, String hashed) {
    return encoder.matches(plain, hashed);
  }

  public String createAccessToken(UserEntity user) {
    Instant now = Instant.now();
    Instant exp = now.plusSeconds(props.getAccessTokenTtlSeconds());
    return JWT.create()
        .withSubject(user.getId())
        .withClaim("nickname", user.getNickname())
        .withClaim("avatarId", user.getAvatarId())
        .withIssuedAt(Date.from(now))
        .withExpiresAt(Date.from(exp))
        .sign(algorithm);
  }

  public DecodeResult decodeAccessToken(String token) {
    try {
      return new DecodeResult(Optional.of(verifier.verify(token).getSubject()), null);
    } catch (TokenExpiredException ex) {
      return new DecodeResult(Optional.empty(), "access_token_expired");
    } catch (JWTVerificationException ex) {
      return new DecodeResult(Optional.empty(), "invalid_token");
    }
  }

  public String generateRefreshToken() {
    byte[] raw = new byte[32];
    SecureRandomHolder.INSTANCE.nextBytes(raw);
    return Base64.getUrlEncoder().withoutPadding().encodeToString(raw);
  }

  public String hashRefreshToken(String raw) {
    try {
      MessageDigest digest = MessageDigest.getInstance("SHA-256");
      byte[] hash = digest.digest(raw.getBytes(StandardCharsets.UTF_8));
      StringBuilder out = new StringBuilder(hash.length * 2);
      for (byte b : hash) {
        out.append(String.format("%02x", b));
      }
      return out.toString();
    } catch (NoSuchAlgorithmException e) {
      throw new IllegalStateException(e);
    }
  }

  public record DecodeResult(Optional<String> subject, String error) {}

  private static final class SecureRandomHolder {
    private static final java.security.SecureRandom INSTANCE = new java.security.SecureRandom();
  }
}

File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/service/SecurityService.kt
package com.hellotimepro.ktor.service

import at.favre.lib.crypto.bcrypt.BCrypt
import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.exceptions.JWTVerificationException
import com.auth0.jwt.exceptions.TokenExpiredException
import com.hellotimepro.ktor.config.AppConfig
import com.hellotimepro.ktor.domain.User
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Instant
import java.util.Base64
import java.util.Date

/** 密码哈希（bcrypt）与 JWT（HS256）+ refresh token 生成/哈希。对应 Spring 的 SecurityService。 */
class SecurityService(private val config: AppConfig) {
    private val algorithm: Algorithm = Algorithm.HMAC256(config.jwtSecret)
    private val verifier = JWT.require(algorithm).build()
    private val secureRandom = SecureRandom()

    fun hashPassword(plain: String): String =
        BCrypt.withDefaults().hashToString(10, plain.toCharArray())

    fun verifyPassword(plain: String, hashed: String): Boolean =
        BCrypt.verifyer().verify(plain.toCharArray(), hashed.toCharArray()).verified

    fun createAccessToken(user: User): String {
        val now = Instant.now()
        return JWT.create()
            .withSubject(user.id.toString())
            .withClaim("nickname", user.nickname)
            .withClaim("avatarId", user.avatarId)
            .withIssuedAt(Date.from(now))
            .withExpiresAt(Date.from(now.plusSeconds(config.accessTokenTtlSeconds)))
            .sign(algorithm)
    }

    /** 校验 access token。过期统一 error="access_token_expired"；其它非法 error="invalid_token"。 */
    fun decodeAccessToken(token: String): DecodeResult = try {
        DecodeResult(verifier.verify(token).subject, null)
    } catch (_: TokenExpiredException) {
        DecodeResult(null, "access_token_expired")
    } catch (_: JWTVerificationException) {
        DecodeResult(null, "invalid_token")
    }

    fun generateRefreshToken(): String {
        val raw = ByteArray(32)
        secureRandom.nextBytes(raw)
        return Base64.getUrlEncoder().withoutPadding().encodeToString(raw)
    }

    fun hashRefreshToken(raw: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(raw.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    data class DecodeResult(val subject: String?, val error: String?)
}
